linera-views/src/shared_view.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(test)]
+#[path = "unit_tests/shared_view.rs"]
+mod tests;
+
+use crate::{
+    batch::Batch,
+    common::Context,
+    views::{RootView, View, ViewError},
+};
+use async_lock::{Semaphore, SemaphoreGuardArc};
+use async_trait::async_trait;
+use futures::channel::oneshot;
+use std::{
+    marker::PhantomData,
+    ops::{Deref, DerefMut},
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc, Mutex,
+    },
+};
+
+#[cfg(not(target_arch = "wasm32"))]
+use crate::{increment_counter, SAVE_VIEW_COUNTER};
+
+/// A way to safely share a [`View`] among multiple readers and at most one writer.
+///
+/// [`View`]s represent some data persisted in storage, but it also contains some state in
+/// memory that caches the storage state and that queues changes to the persisted state to
+/// be sent later. This means that two views referencing the same data in storage may have
+/// state conflicts in memory, and that's why they can't be trivially shared (using
+/// [`Clone`] for example).
+///
+/// The [`SharedView`] provides a way to share an inner [`View`] more safely, by ensuring
+/// that only one writer is staging changes to the view, and than when it is writing those
+/// changes to storage there aren't any more readers for the same view which would have
+/// their internal state become invalid. The readers are not able to see the changes the
+/// writer is staging, and the writer can only save its staged changes after all readers
+/// have finished.
+pub struct SharedView<C, V> {
+    view: V,
+    reader_count: Arc<AtomicUsize>,
+    writer_semaphore: Arc<Semaphore>,
+    writer_notifier: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    _context: PhantomData<C>,
+}
+
+impl<C, V> SharedView<C, V>
+where
+    V: View<C>,
+{
+    /// Wraps a `view` in a [`SharedView`].
+    pub fn new(view: V) -> Self {
+        SharedView {
+            view,
+            reader_count: Arc::new(AtomicUsize::new(0)),
+            writer_semaphore: Arc::new(Semaphore::new(1)),
+            writer_notifier: Arc::new(Mutex::new(None)),
```

Comment:
What would be the issue with using a `Mutex` (instead the semaphore) together with a `RwLock` (instead of the atomic counter and channel)) ?